### ISSUE TITLE
bug in BatchedExecutionEngine::backward: null ndEdfs for late parameter nodes

### ISSUE THREAD

**Issue by mattr1** [clab/dynet#599]
I'm getting null reference errors in BatchedExecutionEngine::backward, in the loop

for (VariableIndex i : cg.parameter_nodes)
    static_cast<ParameterNodeBase*>(cg.nodes[i])->accumulate_grad(ndEdfs[i]);

because ndEdfs[I]->v is null (this happens rarely, but consistently).

I believe this is the cause:

At the end of the BatchedExecutionEngine backward function, *all* of the parameter nodes' backward gradients are accumulated. But, the backward only computes the backward gradients for batches that are computed before whatever batch the "from_where" variable is a part of. 

My final loss is the last node added to the computation graph, but I think with batching it could end up in an earlier batch than some parameter nodes that were added in the same round (between calls to incremental_forward). I don't fully follow the batching code, so let me know if this sounds right. Perhaps this can only happen if those parameter nodes don't affect the final loss?

**Comment by mattr1** [comment]
It seems these are for parameters not used in the calculation of the loss function. Here is a minimal example. The first causes a problem with both SimpleExecutionEngine and BatchedExecutionEngine because it is simply adding a parameter node after the final loss is computed. The second causes a problem only with the BatchedExecutionEngine. It's more subtle: The lookup parameter is added before the loss, but it's batchable, which means it is executed last, and therefore the "loss" batch is lower than the parameter batch, causing problems:

// Causes crash in both SimpleExecutionEngine and BatchedExecutionEngine
void ParameterAfterLossBug() {
	ComputationGraph cg;
	Model model;
	Parameter param = model.add_parameters({ 1 });

	Expression loss = zeroes(cg, { 1 });
	Expression p = parameter(cg, param);

	cg.incremental_forward(loss);
	cg.backward(loss);
}

// Causes crash only in BatchedExecutionEngine
//  - The problem is that the lookup operation is batchable, so is done after all other operations. 
//  - As a result, the finalLoss batch is lower than the lookup batch.
void BatchedParameterAfterLossBug() {
	ComputationGraph cg;
	Model model;
	LookupParameter param = model.add_lookup_parameters(10, { 1 });

	Expression p = lookup(cg, param, 1);
	Expression loss = zeroes(cg, { 1 });

	cg.incremental_forward(loss);
	cg.backward(loss);
}

**Comment by mattr1** [comment]
Thanks!
